### PR TITLE
v1.14 Backports 2023-11-08

### PR DESCRIPTION
**Pull request by aanm**
* [ ] #28810 (@aanm)
 * [ ] #29076 (@aanm)

Once this PR is merged, a GitHub action will update the labels of these PRs:
```upstream-prs
 28810 29076
```

### CODE REVIEW

**Comment by aanm** [comment]
/test-backport-1.14

**Comment by aanm** [comment]
/test-backport-1.14

**Comment by aanm** [comment]
/test-backport-1.14

File: pkg/allocator/allocator.go [hunk 1]
```diff
@@ -579,7 +594,7 @@ func (a *Allocator) lockedAllocate(ctx context.Context, key AllocatorKey) (idpoo
 		return 0, false, false, fmt.Errorf("Found master key after proceeding with new allocation for %s", k)
 	}
 
-	err = a.backend.AllocateIDIfLocked(ctx, id, key, lock)
+	key, err = a.backend.AllocateIDIfLocked(ctx, id, key, lock)
```
Comment (giorio94) [inline pkg/allocator/allocator.go]: Sorry, I missed this during my last review. When `AllocateIDIfLocked` returns an error, the current implementation also returns a nil key. Which means that the error message below will be incorrect.

**Review by squeed — APPROVED** [review]

**Review by giorio94 — CHANGES_REQUESTED** [review]
One comment inline, which I missed during my previous review (and that might be worth addressing).

**Comment by aanm** [comment]
/test-backport-1.14

**Review by giorio94 — APPROVED** [review]

**Pull request merged**